lib/sacastats_web/controllers/character_controller.ex

```diff
@@ -125,6 +137,275 @@ defmodule SacaStatsWeb.CharacterController do
+
+    weapon_best_accuracy =
+      get_highest_rank_weapon(weapons, "weapon_hit_count", "weapon_fire_count")
+
+    weapon_most_kills_per_minute =
+      get_highest_rank_weapon(weapons, "weapon_kills", "weapon_play_time")
+
+    %{
+      "Most Kills" => %{
+        "name" => weapon_most_kills["name"],
+        "type" => "number",
+        "weapon_type" => weapon_most_kills["category"],
+        "image_path" => weapon_most_kills["image_path"],
+        "value" => get_total_values(weapon_most_kills["weapon_kills"])
+      },
+      "Most Killed By" => %{
+        "name" => weapon_most_killed_by["name"],
+        "type" => "number",
+        "weapon_type" => weapon_most_killed_by["category"],
+        "image_path" => weapon_most_killed_by["image_path"],
+        "value" => get_total_values(weapon_most_killed_by["weapon_killed_by"])
+      },
+      "Most Used" => %{
+        "name" => weapon_most_play_time["name"],
+        "type" => "seconds-to-readable",
+        "weapon_type" => weapon_most_play_time["category"],
+        "image_path" => weapon_most_play_time["image_path"],
+        "value" => get_total_values(weapon_most_play_time["weapon_play_time"])
+      },
+      "Most Headshots" => %{
+        "name" => weapon_most_headshots["name"],
+        "type" => "number",
+        "weapon_type" => weapon_most_headshots["category"],
+        "image_path" => weapon_most_headshots["image_path"],
+        "value" => get_total_values(weapon_most_headshots["weapon_headshots"])
+      },
+      "Best Accuracy" => %{
+        "name" => weapon_best_accuracy["name"],
+        "type" => "percentage",
+        "weapon_type" => weapon_best_accuracy["category"],
+        "image_path" => weapon_best_accuracy["image_path"],
+        "value" =>
+          get_percent_ratio(
+            get_total_values(weapon_best_accuracy["weapon_hit_count"]),
+            get_total_values(weapon_best_accuracy["weapon_fire_count"])
+          )
+      },
+      "Best KPM (Kills Per Minute)" => %{
+        "name" => weapon_most_kills_per_minute["name"],
+        "type" => "",
+        "weapon_type" => weapon_most_kills_per_minute["category"],
+        "image_path" => weapon_most_kills_per_minute["image_path"],
+        "value" =>
+          safe_divide(
+            get_total_values(weapon_most_kills_per_minute["weapon_kills"]),
+            trunc(
+              safe_divide(maybe_to_int(weapon_most_kills_per_minute["weapon_play_time"]), 60, 0)
+            ),
+            2
+          )
+      }
+    }
+  end
+
+  def get_highest_rank_weapon(weapons, rank_on)
+      when is_map(weapons) and rank_on === "weapon_play_time" do
+    weapon_list =
+      for {_weapon_id, weapon} <- Map.to_list(weapons) do
+        if weapon["category"] === "Knife" do
+          Map.put(weapon, rank_on, 0)
+        else
+          weapon
+        end
+      end
+
+    Enum.max_by(weapon_list, fn map -> get_total_values(map[rank_on]) end)
```

Comment:
A couple of readability/performance suggestions:
1. Maps are iterable out of the box, so you could say `<- weapons` instead of `<- Map.to_list(weapons)`
2. `for` comprehensions are nice to read, though since you're using it to map values, then find a max, I would recommend a pipeline using `Stream`.
```suggestion
  def get_highest_rank_weapon(weapons, "weapon_play_time") when is_map(weapons) do
    weapons
    |> Stream.map(fn
      {_weapon_id, %{"category" => "Knife"} = weapon} -> Map.put(weapon, rank_on, 0)
      {_weapon_id, weapon} -> weapon
    end)
    |> Enum.max_by(fn map -> get_total_values(map[rank_on]) end)
```
I also turned the conditional logic in the `if` into 2 pattern matching clauses.

The `Stream` module is basically equivalent to `Enum`, except its functions are lazily evaluated. You can think of `Stream.map` as creating a recipe to perform later. In this case, the actual mapping will happen as `Enum.max_by` consumes the element, which results in 1 traversal instead of 2.

As usual please ask questions if you have any :D

Reply:
Will do, thanks!